packages/auth/src/clients/service.ts

```diff
@@ -0,0 +1,153 @@
+import * as crypto from 'crypto'
+import Axios from 'axios'
+import { importJWK, JWK } from 'jose'
+import { URL } from 'url'
+
+import { BaseService } from '../shared/baseService'
+import { IAppConfig } from '../config/app'
+
+interface DisplayInfo {
+  name: string
+  uri: string
+}
+
+interface KeyInfo {
+  proof: string
+  jwk: JWKWithRequired
+}
+
+interface ClientInfo {
+  display: DisplayInfo
+  key: KeyInfo
+}
+
+interface JWKWithRequired extends JWK {
+  kid: string
+  x: string
+}
+
+interface RegistryKey extends JWK {
+  exp?: number
+  nbf?: number
+  revoked?: boolean
+}
+
+interface ServiceDependencies extends BaseService {
+  config: IAppConfig
+}
+
+export interface ClientService {
+  verifySig(
+    sig: string,
+    jwk: JWKWithRequired,
+    challenge: string
+  ): Promise<boolean>
+  validateClientWithRegistry(clientInfo: ClientInfo): Promise<boolean>
+}
+
+export async function createClientService({
+  logger,
+  config
+}: ServiceDependencies): Promise<ClientService> {
+  const log = logger.child({
+    service: 'ClientService'
+  })
+
+  const deps: ServiceDependencies = {
+    logger: log,
+    config
+  }
+
+  return {
+    verifySig: (sig: string, jwk: JWKWithRequired, challenge: string) =>
+      verifySig(deps, sig, jwk, challenge),
+    validateClientWithRegistry: (clientInfo: ClientInfo) =>
+      validateClientWithRegistry(deps, clientInfo)
+  }
+}
+
+async function verifySig(
+  deps: ServiceDependencies,
+  sig: string,
+  jwk: JWKWithRequired,
+  challenge: string
+): Promise<boolean> {
+  const publicKey = (await importJWK(jwk)) as crypto.KeyLike
+  const data = Buffer.from(challenge)
+  return crypto.verify(null, data, publicKey, Buffer.from(sig))
+}
+
+function validateRequiredJwkProperties(jwk: JWKWithRequired): boolean {
+  if (
+    !jwk.kid ||
+    !jwk.x ||
```

Comment:
Wouldn't these be enforced by `JWKWithRequired`?
Should `kid` and `x` instead be checked wherever `clientInfo`'s `key` gets defined?